docs/HowToFAQ/Errors-Refinement.md

```diff
@@ -0,0 +1,981 @@
+The refining declaration needs to be the same kind of declaration as in the base.
+
+
+## Error: an iterator declaration (_name_) in a refining module cannot replace a different kind of declaration in the refinement base** {#ref_iterator_must_refine_iterator}
+
+```dafny
+module P {
+  class I {}
+}
+module Q refines P {
+  iterator I...  {}
+}
+```
+
+Iterators may only refine iterator declarations.
+
+## **Error: a type (_name_) in a refining module may not replace an already defined type (even with the same value)** {#ref_base_type_cannot_be_refined}
+
+<!-- TODO - does not appear to be reachable -->
+
+## **Error: a module (_name_) can only be refined by an alias module or a module facade** {#ref_base_module_must_be_abstract_or_alias}
+
+<!-- TODO  - not sure this is reachable-->
+
+## **Error: a refining iterator is not allowed to add preconditions** {#ref_no_new_iterator_preconditions}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... requires true
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new preconditions may be added
+even if they are implied by the base declarations's preconditions.
+
+## **Error: a refining iterator is not allowed to add yield preconditions** {#ref_no_new_iterator_yield_preconditions}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... yield requires true
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new yield preconditions may be added
+even if they are implied by the base declarations's preconditions.
+
+
+## **Error: a refining iterator is not allowed to extend the reads clause** {#ref_no_new_iterator_reads}
+
+<```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... reads {}
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new reads clauses may be added
+even if they are contained within the base declarations's reads clauses.
+
+
+## Error: a refining iterator is not allowed to extend the modifies clause** {#ref_no_new_iterator_modifies}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... modifies {}
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new modifies clauses may be added
+even if they list objects that are contained within the base declarations's modifies clauses.
+
+
+## **Error: a refining iterator is not allowed to extend the decreases clause** {#ref_no_new_iterator_decreases}
+
+```dafny
+module P {
+  iterator I() yields (x: int)
+}
+module Q refines P {
+  iterator I... decreases 1 {}
+}
+```
+
+There are restrictions on what may change when refining an iterator. In particular, no new decreases clause may be added
+even if if is the same as or implied by the base declaration's decreases clause.
```

Comment:
```suggestion
even if it is the same as or implied by the base declaration's decreases clause.
```

Reply:
Nit: above you use plural "modifies clause_s_", here you use singular "decreases clause".